x/authenticator/keeper/msg_server.go

```diff
@@ -19,21 +23,57 @@ func NewMsgServerImpl(keeper Keeper) types.MsgServer {
 
 var _ types.MsgServer = msgServer{}
 
+// AddAuthenticator allows the addition of various types of authenticators to an account.
+// This method serves as a versatile function for adding diverse authenticator types
+// to an account, making it highly adaptable for different use cases.
+
 func (m msgServer) AddAuthenticator(
 	goCtx context.Context,
 	msg *types.MsgAddAuthenticator,
 ) (*types.MsgAddAuthenticatorResponse, error) {
 	ctx := sdk.UnwrapSDKContext(goCtx)
 
-	// TODO: validate based on type of authenticator
-	// TODO: do we want to create and account for the data/pubkey
-	// here? Or in ante handler
-
 	sender, err := sdk.AccAddressFromBech32(msg.Sender)
 	if err != nil {
 		return nil, err
 	}
 
+	authenticators, err := m.Keeper.GetAuthenticatorsForAccount(ctx, sender)
+	if err != nil {
+		return nil, err
+	}
+
+	// If there are no other authenticators, ensure that the first authenticator is a SignatureVerificationAuthenticator.
+	if len(authenticators) == 0 && msg.Type != authenticator.SignatureVerificationAuthenticatorType {
+		return nil, fmt.Errorf("the first authenticator must be a SignatureVerificationAuthenticator")
+	}
+
+	// Validate that the data is correct for each type of authenticator.
+	switch msg.Type {
+	case authenticator.SignatureVerificationAuthenticatorType:
```

Comment:
Like we discussed in the call. This could be replaced by: `authenticator.Validate(account, msg.Data)` or  `authenticator.OnAuthenticatorAdded(account, msg.Data)`. But that doesn't need to block this PR, we can just add it on top

Reply:
I'll add that next :muscle: